Dockerfile

```diff
@@ -28,11 +28,14 @@ ADD . /server
 # Configure and build
 RUN mkdir docker_build && cd docker_build && cmake .. && make -j $(nproc)  && cd .. && rm -r /server/docker_build
 
-# Copy the docker config files to the conf folder instead of the default config
-COPY /conf/default/* conf/
-
-# Copy the docker settings files to the settings folder instead of the default settings
-COPY /scripts/settings/default/* scripts/settings/
+# Set SQL_HOST in lua settings 
+RUN echo '\
```

Comment:
I think a better way of doing this is to:

- Still COPY the default settings up one level
- Create a new script in tools that replaces the contents of `settings/network.lua` with whatever else it is that you need

If memory serves, `>>` is append and `>` is replace. These files are designed to only have a single table with relevant keys in. If it's working with multiple tables in a file, that's entirely not by design and I wouldn't rely on that.

Reply:
I've been looking at this in my own branch, and I'm trying to debug some stuff but I think we might be over-complicating this a little bit.

The docker file itself shouldn't have any default settings in it, it's supposed to be configuration agnostic. The settings should be configured in the compose file.

That said, let's say my compose file has env vars: XI_NETWORK_SQL_HOST, XI_NETWORK_SQL_PORT.

In the settings.ccp where it does defaults -> user settings, the last step should be applying any environment variables over the user settings: 

if(getenv(XI_{}_{}), outerKey, innerKey)
  // Apply the environment setting.
  
But for some reason in my branch, even though I can verify the settingsMap is getting updated, there's still some issues.

This seems like the correct approach though because it keeps the docker concerns (build, configure) separated, and it will work with any future config values added so you wouldn't have to touch settings.cpp again ever.

Reply:
> In the settings.ccp where it does defaults -> user settings, the last step should be applying any environment variables over the user settings:
> if(getenv(XI_{}_{}), outerKey, innerKey)
> // Apply the environment setting.

+1 for this approach, then we can get rid of those hard-coded function blocks that try to read the env settings in

Reply:
That would maybe look like this:
```cpp
        // Scrape user settings into cpp's settingsMap
        // This will overwrite the defaults, if user settings exist. Otherwise the
        // defaults will be left intact.
        for (auto [outerKeyObj, outerValObj] : lua["xi"]["settings"].get<sol::table>())
        {
            auto outerKey = outerKeyObj.as<std::string>();

            for (auto [innerKeyObj, innerValObj] : outerValObj.as<sol::table>())
            {
                auto innerKey = innerKeyObj.as<std::string>();
                auto key      = to_upper(fmt::format("{}.{}", outerKey, innerKey));

                if (innerValObj.is<bool>())
                {
                    settingsMap[key] = innerValObj.as<bool>();
                }
                else if (innerValObj.is<double>())
                {
                    settingsMap[key] = innerValObj.as<double>();
                }
                else if (innerValObj.is<std::string>())
                {
                    settingsMap[key] = innerValObj.as<std::string>();
                }

                auto envKey = fmt::format("XI_{}_{}", to_upper(outerKey), to_upper(innerKey));
                if (auto* value = std::getenv(envKey.c_str()))
                {
                    settingsMap[key] = value;
                }
            }
        }
```

Reply:
Actually, we probably want a little logging here too:
```cpp
                auto envKey = fmt::format("XI_{}_{}", to_upper(outerKey), to_upper(innerKey));
                if (auto* value = std::getenv(envKey.c_str()))
                {
                    ShowInfo(fmt::format("Applying setting from environment variable {} -> {}", key, value).c_str());
                    settingsMap[key] = value;
                }
```

Reply:
I'll try this in my branch when I get home, but it's more or less what I was doing (I was just doing it in a separate loop, similar to how default/user settings were scraped).

But I get the feeling there's something weird going on with how the settings bubble back up to the lua[] config from the settingsMap that makes it not end up with the right values, so I was planning on just populating the lua again, and putting in the map last. Either way this looks cleaner and I'll try and test it out a bit when I get a chance.

Reply:
> Actually, we probably want a little logging here too:
> 
> ```c++
>                 auto envKey = fmt::format("XI_{}_{}", to_upper(outerKey), to_upper(innerKey));
>                 if (auto* value = std::getenv(envKey.c_str()))
>                 {
>                     ShowInfo(fmt::format("Applying setting from environment variable {}: {} -> {}", envKey, key, value).c_str());
>                     settingsMap[key] = value;
>                 }
> ```

@zach2good Hey, sorry only got to this now and figured I'd ask here. I'm not a C++ guru, haven't used it in like 10 years so a lot of the 14/17 features are new to me.

After implementing this idea, I'm getting the same behavior as I was when I tried it my way: calling 
`get<std::string>("NETWORK.SQL_HOST)`
for instance, prints "TRUE", so it's storing the values as booleans not strings. Any idea what might be going on there between the auto and variant<bool, double, string> types? Thanks.

Once I get this working, I'll make a base fork + write up a PR for a wiki page after ironing out the other kinks.